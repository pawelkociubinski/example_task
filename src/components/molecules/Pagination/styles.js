import styled, { css } from 'styled-components';

export const Container = styled.div`
  align-items: center;
  display: flex;
  justify-content: center;
`;

export const PreviousPage = styled.div`
  cursor: pointer;
  margin-right: 30px;
`;

export const NextPage = styled.div`
  cursor: pointer;
  margin-left: 30px;
`;

export const Dot = styled.div`
  display: flex;
  ${({ showEllipsis }) => {
    if (showEllipsis) {
      return css`
        & + & {
          display: none;
        }
      `;
    }
  }};
`;
