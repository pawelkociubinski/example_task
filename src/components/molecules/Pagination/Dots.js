/* @flow */
// Dependencies
import React, { Component } from 'react';
// Custom
import type { ChangePageEvent } from 'types/actions';

import Button from '../../atoms/Button';
import { Dot } from './styles';

type Props = {
  currentPage: number,
  page: number,
  changePage: (page: number) => ChangePageEvent,
  distanceToCurrentPage: number,
  edgePage: boolean,
  isAboveRangeLimit: boolean,
  preservePages: number,
  totalPages: number,
}

class Dots extends Component<Props> {
  get isActive() {
    const { currentPage, page } = this.props;

    return page === currentPage;
  }

  _changePage = (): void => {
    const { changePage, page } = this.props;

    changePage(page);
  };

  render() {
    const {
      currentPage,
      distanceToCurrentPage,
      edgePage,
      isAboveRangeLimit,
      page,
      preservePages,
      totalPages,
    } = this.props;

    const isNeighborOfCurrentPage = distanceToCurrentPage > 1;
    const isStartPreserved =
      page <= preservePages && currentPage < preservePages;
    const isEndPreserved =
      page >= totalPages - preservePages &&
      currentPage > totalPages - preservePages;

    const displayEllipsis =
      isAboveRangeLimit &&
      isNeighborOfCurrentPage &&
      !edgePage &&
      !isStartPreserved &&
      !isEndPreserved;

    return (
      <Dot showEllipsis={displayEllipsis}>
        <Button
          shape="circle"
          handleClick={this._changePage}
          theme={this.isActive ? 'primary' : 'transparent'}
        >
          {displayEllipsis ? '...' : page + 1}
        </Button>
      </Dot>
    );
  }
}

export default Dots;
