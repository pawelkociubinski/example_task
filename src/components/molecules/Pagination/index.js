/* @flow */
// Dependencies
import React, { Component } from 'react';
// Custom
import type { PreviousPageEvent, NextPageEvent, ChangePageEvent } from 'types/actions';

import Icon from '../../atoms/Icon';
import Dots from './Dots';
import { Container, NextPage, PreviousPage } from './styles';

type Props = {
  itemsLength: number,
  itemsPerPage: number,
  currentPage: number,
  previousPage: () => PreviousPageEvent,
  nextPage: () => NextPageEvent,
  changePage: (page: number) => ChangePageEvent,
};

type State = {
  rangeOfPages: number,
  preservePages: number,
}

class Pagination extends Component<Props, State> {
  constructor() {
    super();
    this.state = {
      rangeOfPages: 6,
      preservePages: 4,
    };
  }

  get itemsOnPage() {
    const { itemsLength, itemsPerPage } = this.props;

    return Math.ceil(itemsLength / itemsPerPage) - 1;
  }

  get totalPages() {
    return this.itemsOnPage >= 0 ? this.itemsOnPage : 0;
  }

  get isAboveRangeLimit() {
    const { rangeOfPages } = this.state;

    return this.totalPages + 1 >= rangeOfPages;
  }

  get pages() {
    return [...Array(this.totalPages + 1).keys()];
  }

  render() {
    const {
      previousPage, nextPage, changePage, currentPage,
    } = this.props;
    const { preservePages } = this.state;

    const previousPageProps = {
      onClick: currentPage !== 0 ? previousPage : undefined,
    };

    const nextPageProps = {
      onClick: currentPage !== this.totalPages ? nextPage : undefined,
    };
    return (
      <Container>
        <PreviousPage {...previousPageProps}>
          <Icon size={16} icon="arrowLeft" />
        </PreviousPage>
        {this.pages.map((page) => {
          const distanceToCurrentPage = Math.abs(page - currentPage);
          const edgePage = page === this.totalPages || page === 0;

          return (
            <Dots
              changePage={changePage}
              currentPage={currentPage}
              distanceToCurrentPage={distanceToCurrentPage}
              edgePage={edgePage}
              isAboveRangeLimit={this.isAboveRangeLimit}
              key={page}
              page={page}
              preservePages={preservePages}
              totalPages={this.totalPages}
            />
          );
        })}
        <NextPage {...nextPageProps}>
          <Icon size={16} icon="arrowRight" />
        </NextPage>
      </Container>
    );
  }
}

export default Pagination;
