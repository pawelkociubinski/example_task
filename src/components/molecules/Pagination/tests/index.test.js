import React from 'react';
import renderer from 'react-test-renderer';
import Pagination from '../index';

describe('<Pagination />', () => {
  it('renders correctly', () => {
    const mockedFn = jest.fn();
    const currentPage = 0;
    const repositoriesLength = 10;
    const itemsPerPage = 2;

    const component = renderer.create(<Pagination
      changePage={mockedFn}
      currentPage={currentPage}
      itemsLength={repositoriesLength}
      itemsPerPage={itemsPerPage}
      nextPage={mockedFn}
      previousPage={mockedFn}
    />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
