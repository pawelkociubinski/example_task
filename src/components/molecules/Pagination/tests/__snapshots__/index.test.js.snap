// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders correctly 1`] = `
.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #DDDDDD;
  color: #707070;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16px;
  height: 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  overflow: hidden;
  position: relative;
  border: none;
  border-radius: 24px;
  padding: 0;
  width: 48px;
}

.c3:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  color: #707070;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16px;
  height: 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  overflow: hidden;
  position: relative;
  border: none;
  border-radius: 24px;
  padding: 0;
  width: 48px;
}

.c4:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  cursor: pointer;
  margin-right: 30px;
}

.c5 {
  cursor: pointer;
  margin-left: 30px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={undefined}
  >
    <svg
      className=""
      height={16}
      viewBox="0 0 40 40"
      width={16}
    >
      <path
        d="m26.5 12.1q0 0.3-0.2 0.6l-8.8 8.7 8.8 8.8q0.2 0.2 0.2 0.5t-0.2 0.5l-1.1 1.1q-0.3 0.3-0.6 0.3t-0.5-0.3l-10.4-10.4q-0.2-0.2-0.2-0.5t0.2-0.5l10.4-10.4q0.3-0.2 0.5-0.2t0.6 0.2l1.1 1.1q0.2 0.2 0.2 0.5z"
      />
    </svg>
  </div>
  <div
    className="c2"
  >
    <button
      className="c3"
      onClick={[Function]}
    >
      1
    </button>
  </div>
  <div
    className="c2"
  >
    <button
      className="c4"
      onClick={[Function]}
    >
      2
    </button>
  </div>
  <div
    className="c2"
  >
    <button
      className="c4"
      onClick={[Function]}
    >
      3
    </button>
  </div>
  <div
    className="c2"
  >
    <button
      className="c4"
      onClick={[Function]}
    >
      4
    </button>
  </div>
  <div
    className="c2"
  >
    <button
      className="c4"
      onClick={[Function]}
    >
      5
    </button>
  </div>
  <div
    className="c5"
    onClick={[MockFunction]}
  >
    <svg
      className=""
      height={16}
      viewBox="0 0 40 40"
      width={16}
    >
      <path
        d="m26.3 21.4q0 0.3-0.2 0.5l-10.4 10.4q-0.3 0.3-0.6 0.3t-0.5-0.3l-1.1-1.1q-0.2-0.2-0.2-0.5t0.2-0.5l8.8-8.8-8.8-8.7q-0.2-0.3-0.2-0.6t0.2-0.5l1.1-1.1q0.3-0.2 0.5-0.2t0.6 0.2l10.4 10.4q0.2 0.2 0.2 0.5z"
      />
    </svg>
  </div>
</div>
`;
